net-p2p/mldonkey/mldonkey-3.1.7.ebuild

```diff
@@ -42,7 +42,7 @@ DEPEND="${RDEPEND}
 	<dev-lang/ocaml-4.10:=[ocamlopt?]
 	bittorrent? (
 		|| (
-			>=dev-lang/ocaml-4.06[ocamlopt?]
+			<dev-lang/ocaml-4.06[ocamlopt?]
 			dev-ml/num
```

Comment:
This doesn't seem to be working. If I install dev-ml/num, this compilation fails.
```
File "src/utils/extlib/IO.ml", line 70, characters 22-23:
70 | 			let r = i.in_input s !p !l in
                           ^
Error: This expression has type bytes but an expression was expected of type
         string
/usr/bin/ocamlc.opt -w -3  -I src/utils/extlib  -I src/utils/cdk  -I src/utils/bitstring  -I src/utils/lib  -I src/utils/ocamlrss  -I src/utils/xml-light  -I src/utils/net  -I tools  -I src/daemon/common  -I src/daemon/driver  -I src/utils/mp3tagui  -I src/config/unix  -I src/gtk2/gui  -I src/gtk2/gui/x11  -I src/gtk2/gui/win32  -I icons/rsvg  -I +lablgtk2  -I src/networks/direct_connect  -I src/networks/fileTP  -I src/networks/bittorrent  -I src/networks/donkey -I +camlp4 -c src/utils/lib/store.mli
make: *** [Makefile:5444: src/utils/extlib/IO.cmx] Error 2
make: *** Waiting for unfinished jobs....
 * ERROR: net-p2p/mldonkey-3.1.6-r1::testworld failed (compile phase):
 *   emake failed
```
And I can't install `<dev-lang/ocaml-4.06[ocamlopt?]`, portage just won't allow me to. Probably some other deps require higher version?

```
dev-ml/camlp4-4.11_p1 (>=dev-lang/ocaml-4.11.0[ocamlopt?])
dev-ml/findlib-1.8.1-r1 (>=dev-lang/ocaml-4.08.1-r1[ocamlopt?])
dev-ml/num-1.3-r1 (>=dev-lang/ocaml-4.09.0[ocamlopt])

```
You maybe have to limit camlp4 version that doesn't require as high ocaml? But then again this is starting to block many ocaml-related updates.

Maybe we just mask bittorrent use for this version? Would it be a huge loss, does most people use the functionality?

Reply:
But there you are building net-p2p/mldonkey-3.1.6-r1.
Are we aiming at fixing that version too? I though the plan was stabilizing net-p2p/mldonkey-3.1.7 to remove older versions that are affected by many bugs, some of them unsolvable.

Reply:
Okay, now it works, but if I let portage try to emerge this on itself, it will prefer to 3.1.6-r1. @thesamesam do you happen to have ideas why? Is it fine?

Now this version builds, LGTM. Please go ahead and merge if you find this okay.

Reply:
Usually when that happens, trying to emerge the version you _know_ is out there is helpful. For me, I had:
```
These are the packages that would be merged, in order:

Calculating dependencies... done!
[binary     UD ] sys-libs/binutils-libs-2.34-r2:0/2.34::gentoo [2.35.1:0/2.35.1::gentoo] USE="-64-bit-bfd -multitarget -nls -static-libs" ABI_X86="(64) -32 (-x32)" 0 KiB
[binary     UD ] dev-lang/ocaml-4.09.0:0/4.09.0::gentoo [4.11.1:0/4.11.1::gentoo] USE="ocamlopt -emacs -flambda -latex -spacetime -xemacs" 0 KiB
[binary  N     ] dev-ml/camlp4-4.11_p1:0/4.11_p1::gentoo  USE="ocamlopt" 0 KiB
[ebuild  N     ] net-p2p/mldonkey-3.1.7::gentoo  USE="ocamlopt -bittorrent -doc -fasttrack -gd -gnutella -gtk -guionly -magic -upnp" 0 KiB

Total: 4 packages (2 downgrades, 2 new, 3 binaries), Size of downloads: 0 KiB

!!! Multiple package instances within a single package slot have been pulled
!!! into the dependency graph, resulting in a slot conflict:

dev-lang/ocaml:0

  (dev-lang/ocaml-4.09.0:0/4.09.0::gentoo, binary scheduled for merge) USE="ocamlopt -emacs -flambda -latex -spacetime -xemacs" ABI_X86="(64)" pulled in by
    <dev-lang/ocaml-4.10:=[ocamlopt?] required by (net-p2p/mldonkey-3.1.7:0/0::gentoo, ebuild scheduled for merge) USE="ocamlopt -bittorrent -doc -fasttrack -gd -gnutella -gtk -guionly -magic -upnp" ABI_X86="(64)"
    ^               ^^^^ ^

  (dev-lang/ocaml-4.11.1:0/4.11.1::gentoo, installed) USE="ocamlopt -emacs -flambda -latex -spacetime -xemacs" ABI_X86="(64)" pulled in by
    >=dev-lang/ocaml-4.11.0:0/4.11.1=[ocamlopt] required by (dev-ml/camlp4-4.11_p1:0/4.11_p1::gentoo, binary scheduled for merge) USE="ocamlopt" ABI_X86="(64)"
    ^^               ^^^^^^^^^^^^^^^^
    >=dev-lang/ocaml-4.02.3-r1:0/4.11.1=[ocamlopt] required by (dev-ml/ocamlbuild-0.14.0-r1:0/0.14.0::gentoo, installed) USE="ocamlopt -test" ABI_X86="(64)"
                              ^^^^^^^^^^

sys-libs/binutils-libs:0

  (sys-libs/binutils-libs-2.34-r2:0/2.34::gentoo, binary scheduled for merge) USE="-64-bit-bfd -multitarget -nls -static-libs" ABI_X86="(64) -32 (-x32)" pulled in by
    sys-libs/binutils-libs:0/2.34= required by (dev-lang/ocaml-4.09.0:0/4.09.0::gentoo, binary scheduled for merge) USE="ocamlopt -emacs -flambda -latex -spacetime -xemacs" ABI_X86="(64)"
                          ^^^^^^^^

  (sys-libs/binutils-libs-2.35.1:0/2.35.1::gentoo, installed) USE="-64-bit-bfd -multitarget -nls -static-libs" ABI_X86="(64) -32 (-x32)" pulled in by
    sys-libs/binutils-libs:0/2.35.1= required by (dev-lang/ocaml-4.11.1:0/4.11.1::gentoo, installed) USE="ocamlopt -emacs -flambda -latex -spacetime -xemacs" ABI_X86="(64)"
                          ^^^^^^^^^^
    (and 1 more with the same problem)

NOTE: Use the '--verbose-conflicts' option to display parents omitted above
```

So, I think it's harmless, just a consequence of having some newer OCaml things with tight deps. Unfortunately, we're going to be in that land for a little bit longer until applications have ported properly to use safe strings. :/